### PR TITLE
NPCs reloading mags

### PR DESCRIPTION
**Pull request by Coolthulhu**
Has some shortcomings:
- Goes over the items multiple times. If NPCs had big inventories, it would be slow
- Uses high-level functions to stay consistent with player methods. High level means slow.
- Casting away `const`s in some places, because `item_location` requires a non-`const` `Character`. Any good ideas?
- NPCs don't understand `RELOAD_AND_SHOOT`. They will reload bows, slings etc. This doesn't seem to cause serious problems, though.
- Some problems with priorities - NPCs will sometimes make dumb decisions because they don't understand how slow some actions are
- They will reload auxiliary gunmods, but with unpredictable priority (only when wielded?) and won't use them.

Other than that, it seems to work. NPCs can put mags into guns, reload mags, reload mags inside guns and reload flamethrowers.

Also fixed NPCs not switching weapons.

### CODE REVIEW

**Comment by mugling** [comment]
> Casting away consts in some places, because item_location requires a non-const Character. Any good ideas?

Unfortunately, no. I also came across this problem also when refactoring `item_location` and there wasn't an obvious solution. This is reason why `find_ammo` and `pick_reload_ammo` are both non-`const`.

On the hand the constructor should require a mutable `Character&` as the class is quite capable of mutating it but in others it's awkward because you can't construct a `const item_location` using `const Character&`.

I did look at doing a `const_cast` inside the class itself but that seems more evil than specific instances where it is otherwise unavoidable.

File: src/item.cpp [hunk 1]
```diff
@@ -4337,6 +4338,11 @@ item::reload_option item::pick_reload_ammo( player &u ) const
         return ( lhs.ammo->ammo_remaining() != 0 ) > ( rhs.ammo->ammo_remaining() != 0 );
     } );
 
+    // NPCs always use the first option
```
Comment (mugling) [inline src/item.cpp]: Arguably a redundant comment as the code below is unambiguous and self documenting

**Comment by Coolthulhu** [comment]
> On the hand the constructor should require a mutable Character& as the class is quite capable of mutating it but in others it's awkward because you can't construct a const item_location using const Character&.

Construct it using factory methods maybe?

File: src/npcmove.cpp [hunk 2]
```diff
@@ -931,31 +909,91 @@ npc_action npc::address_needs()
...
 
-bool npc::can_reload()
+bool wants_to_reload( const item &it )
 {
-    return find_usable_ammo();
+    if( !it.can_reload() ) {
+        return false;
+    }
+
+    const int required = it.ammo_required();
```
Comment (mugling) [inline src/npcmove.cpp]: consider `item::is_reloadable()` and/or `item::can_reload()` as these will be updated when bandoliers become directly reloadable.

File: src/npcmove.cpp [hunk 3]
```diff
@@ -931,31 +909,91 @@ npc_action npc::address_needs()
...
+
+    return ret_null;
+}
+
+const item &npc::find_reloadable() const
+{
+    return const_cast<const item &>( const_cast<npc *>( this )->find_reloadable() );
+}
+
+bool npc::can_reload()
```
Comment (mugling) [inline src/npcmove.cpp]: potential for confusion with `item::can_reload`?

File: src/npcmove.cpp [hunk 4]
```diff
@@ -2818,3 +2865,30 @@ bool npc::complain()
...
+void npc::do_reload( item &it )
+{
+    auto usable_ammo = find_usable_ammo( it );
+
+    if( !usable_ammo ) {
+        debugmsg( "do_reload failed: no usable ammo" );
+        return;
+    }
+
+    int reload_time = item_reload_cost( it, *usable_ammo );
```
Comment (mugling) [inline src/npcmove.cpp]: you need to pass `qty` here to get the correct (reduced) times for partially loaded items

**Comment by mugling** [comment]
> Construct it using factory methods maybe?

As in hide a `const_cast` inside a static member function?

```
const item_location item_location::on_character( const Character &ch, item *which ) {
    return item_location( const_cast<Character&>( ch, which ) );
}
```

**Comment by Coolthulhu** [comment]
Changed the `npc::can_reload` to `npc::can_reload_current` and added the quantity to reload parameters.

Didn't fix the const casting.

**Comment by mugling** [comment]
> Didn't fix the const casting.

The implementation you have is probably the simplest for the time being. If such casts are excessive we can work out a better solution then.

File: src/player.h [hunk 5]
```diff
@@ -747,7 +747,7 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
          * @param prompt optional message to display in any menu
          * @return whether the item was successfully disposed of
          */
-        bool dispose_item( item& obj, const std::string& prompt = std::string() );
+        virtual bool dispose_item( item& obj, const std::string& prompt = std::string() );
```
Comment (mugling) [inline src/player.h]: eventually we should probably push this into `Character`

**Comment by mugling** [comment]
> Some problems with priorities

NPC's often reload too frequently (in some circumstances after each shot)

**Pull request merged**